### DISCUSSION TITLE
Integrating in build process

### DISCUSSION THREAD

**Discussion by hylkevds** [javacc21/javacc21#179] [Q&A]
The new features of JavaCC 21 (CongoCC?) sound really promising, especially the Context-Sensitive Tokenizing!
They would solve quite a few difficulties I currently have with the original JavaCC.

In my current setup I use the JavaCC maven plugin to generate the Parser classes during the build of my project, meaning only the grammar files are in the source code repository. Is there a way to do that with JavaCC 21, or is one expected to generate the parser outside of the build, and put the generated Parser code in git?

**Comment by revusky** [comment]
Well, I certainly wouldn't suggest that you commit any *generated* parser code to the git repository. You should surely just have the grammar file in git and everything else should be generated as part of the build. But generally speaking, this is not really any sort of novel problem. All you need is to check whether the grammar file is newer than the generated parser file and, if so, regenerate the parser code.

This project is using Ant internally, not Maven, but surely the principle is about the same. For example, the internal Java grammar has its own [Ant build script](https://github.com/javacc21/javacc21/blob/master/examples/java/build.xml) for building/testing the Java grammar on its own. That build script is quite trivial, the key part  (from your POV) being [here](https://github.com/javacc21/javacc21/blob/master/examples/java/build.xml#L9-L20). That's ant, but I can't believe this is hard to do with maven. 

I hope that answers your question. By the way, if you have any other questions, it would be preferable to use the [discussion forum here](https://discuss.congocc.org/). That is currently more active than this and I may just disable this discussion feature pretty soon.

**Comment by hylkevds** [comment]
But how would the build script access the JavaCC 21 jar to do the generation, if JavaCC 21 is not available in Maven central (or any other maven repository)?
It seems to me that putting the JavaCC21 jar into Git is a much worse solution that putting the generated source code in Git.

The legacy JavaCC maven plugin nicely integrates into the maven build process. Are there examples on how to do that with JavaCC 21?

**Comment by revusky** [comment]
Well, I would point out that you're able to run things like `javac` and `jar` and those aren't in Maven Central. (They're part of the JDK.) So, somehow, you are able to run some programs without them being on Maven Central.

But anyway...

> It seems to me that putting the JavaCC21 jar into Git is a much worse solution that putting the generated source code in Git.

Well, I can't really agree with that. Putting generated source code in your version control repository is really a pretty terrible idea and strikes me as *much* worse than putting a jarfile (or a few of them) in there. In fact, having a bin or libs directory in your project that just contains the necessary binaries to bootstrap your build does not strike me as a particularly *evil* practice. (If it is evil, I'm missing something about all this!) And, frankly, that likely is the simplest solution to your problem.

But, okay, if you want to avoid that, then you can grab the latest javacc-full.jar from the server as part of a "clean" build, that can be picked up at https://javacc.com/download/javacc-full.jar and it should not be hard to integrate that into your build process.

Ant has a [task called 'get'](https://ant.apache.org/manual/Tasks/get.html) which allows you to grab a file from a URL (with various options like up-to-date check and checksum and so forth...) and one would think that there is the same thing for maven. I was going to google it up and tell you as a courtesy, but I can't find such a thing and it seems you need a separate plugin to download a file from somewhere. See: https://github.com/maven-download-plugin/maven-download-plugin

So, I guess, with that, you can just download the javacc-full.jar jarfile (if necessary) and thus, there would be no need for it to be checked into git. I think you ought to be able to sort through this. I guess, at some point, probably after the CongoCC renaming, we should put together some sort of CongoCC maven plugin (since it seems to be important for some people.) But meanwhile, I feel I should say pretty bluntly that I'm much more motivated to answer questions about usage of the tool itself than these sorts of integration questions that really aren't our problem! I'm a bit retro myself and have never used maven or gradle either, so I don't really know the details about such questions. All I know is that it is surely not a very difficult problem!

**Comment by hylkevds** [comment]
Thank you for the reply.

> Well, I would point out that you're able to run things like javac and jar and those aren't in Maven Central. (They're part of the JDK.) So, somehow, you are able to run some programs without them being on Maven Central.

Yes, by all means, if you get CongoCC accepted as Debian package, like OpenJDK and Maven are, that works for me too. 
:)
Though I fear they'll reject it based on the lack of versioning. :(

> (since it seems to be important for some people.)

I think you are massively underestimating the size of this "some" population. It's about 99.9% of software developers.

I'll look into the various options. I guess putting the jar in GIT is the most reliable, since that won't cause the build to break if you make an accidental incompatible change.

**Comment by revusky** [comment]
Well, currently the project is in a bit of a transitional state, going from the JavaCC (21) naming to CongoCC. So, I don't see making any effort to get the tool into various channels like Debian etc. until that transition is taken care of. And then it really becomes a question of how to get more projection for the project, I guess. 

But, you know, I do have to say that you may be exaggerating the real value of these sorts of things. For example, out of curiosity, I just typed:

      sudo apt install javacc

on Ubuntu and it installed fine. The legacy JavaCC, obviously. And you can then type `javacc` and `jjtree` on the command line, but these are quite old versions, it says 5.0 and I think that is what their release number was well over 10 years ago. Now, it so happens that, even though they are now on 7.0.11, and over a decade has passed, there is basically not a single new feature in all of those version number changes, so it doesn't matter anyway! But, what I mean to say is that having something in a channel like that, well, if there really is development, somebody has to keep it up to date! (I don't think there's that much issue with versioning. One can just use the date of the snapshot as a version number, so something from today could be 22.05.17 or something.)

With  Debian in particular, a lot of the packages in their channel are very out of date. Now, I know that they say that this is a great thing, part of their "philosophy" because they're so into "stability" but frankly.... c'mon... The problem remains that if nobody takes ownership of the problem of keeping these things up to date, then the whole thing is of rather dubious value. I mean, it's still true that it's super easy to install the tool, some version of it anyway, but I don't think that having all these stale versions of something floating around on Maven Central or anywhere else is necessarily so great. But again, it does relate to whether there is somebody committed to doing the minimal maintenance around these things.

With these very high profile projects, like the JDK, sure, there's surely somebody very committed to keeping it up to date in the various channels, I suppose. And the people who maintain Linux distros themselves are motivated to keep something important like that up to date, but not something like this.

And meanwhile, it's just not very hard to "install" something like JavaCC 21. It's just a totally self-contained jarfile. It's just: 

       java -jar javacc-full.jar MyGrammar

and if you want to have a bash script or .bat file in your PATH that invokes that, then that's not hard either.

But understand, it's not that I'm against all of this so much either. Certainly, if you or anybody else wants to take ownership of the problem, like write a Maven plugin that grabs the latest jar and so on, then I'm happy, but you know, then there is the issue of whether you (or whoever) are committing to keeping the thing working. But making it easy to start using something is a good thing, sure.

Anyway, I see that you wrote something on the other discussion forum, so any conversation can continue there. I'm probably going to disable this discussions feature here at some point soon, because there's not much reason to have discussion forums all over the place.

**Comment by hylkevds** [comment]
I totally agree that getting the name change done is the first priority.
I'm experimenting a bit, and will post more after that. Hopefully with a tutorial for other beginners :)

**Comment by vsajip** [comment]
> I'm probably going to disable this discussions feature here at some point soon, because there's not much reason to have discussion forums all over the place.

I think there's still some value in keeping the discussion forum here for source code-related questions / discussions, because of the ease of e.g. pointing to code snippets in the current source, at least for collaborators if not the wider user community.

> you can grab the latest javacc-full.jar from the server as part of a "clean" build, that can be picked up at https://javacc.com/download/javacc-full.jar and it should not be hard to integrate that into your build process.

The problem I see here is versioning. A build process in the corporate environment which doesn't allow you to specify a particular version of a jar to be downloaded will run into all sorts of problems, both technical and organizational. This is for good reasons, IMO, as regressions are always possible and one often has to download a specific older version of a tool or library to not break a build. It would become easier for CongoCC to gain wider acceptance if we adopt a more conventional versioning scheme, as the project matures further.

I personally see little value in uploading CongoCC to Maven Central, other than to satisfy corporate guidelines mandated in some enviroments. ISTM Maven Central is more for libraries rather than applications.

**Reply by revusky** [reply]
> The problem I see here is versioning. A build process in the corporate environment which doesn't allow you to specify a particular version of a jar to be downloaded will run into all sorts of problems, both technical and organizational. This is for good reasons, IMO, as regressions are always possible and one often has to download a specific older version of a tool or library to not break a build. 

Well, as far as I can see, the way to deal with this sort of problem is just to check in the version of the jarfile that you're using into version control. And then, periodically, when you are satisfied that your code builds and runs with an updated version, then update the checked-in jarfile. But if you do that, you still have the history of the jarfiles you were using at any previous point in time, so any checkout of an earlier snapshot of the code will grab the jarfile being used at that point in time and the build will be reproducible. (Though, in that scenario, it should also be of some interest to see whether the code builds and runs with the latest javacc21 jarfile, but I guess that is a separate build target finally.) 

> I personally see little value in uploading CongoCC to Maven Central, other than to satisfy corporate guidelines mandated in some environments. 

Well, this is it, I guess. I'm not so opposed to jumping through a hoop or two to make some corporate people happy. However, I am disinclined to pretend that it really means that much! Most of this is just formalistic really. At least in our context, it is. But okay, if this is so important to certain people, fine, but the problem right now is that there are so many things to do to get the CongoCC transition going that we should just concentrate on that over the next few months and then maybe take a breather and do some less fundamental things, like thinking about overcoming whatever hangups certain corporate people have.

Though, I also don't believe that this stuff is what is causing slow adoption. I think it's mostly that people don't know about it! So we need to seriously think about how to the word out, sure.

> ISTM Maven Central is more for libraries rather than applications.

Well, I'm pretty satisfied that, in our context, all this Maven Central stuff is not really solving any problem that this project has. What's more, just philosophically, I am (probably like you!) rather unenthusiastic about having all these stale versions of the tool floating around there or anywhere else. As best I can see, once you have a high degree of confidence that your code builds and runs correctly with the *current* version of the tool, there is really just about no practical reason to ever use anything older than that. (Maybe you can construct some scenarios where there is some reason to build/test against obsolete versions, but I would say it is very rare.)

When you have a self-contained tool that has no dependencies aside from the JDK. And then, in turn, the code that the tool generates also has no dependencies aside from the JDK... I still hold the view that, in terms of the developer's typical workflow, I mean the edit->rebuild->retest cycle, I really think our stance [here](https://doku.javacc.com/doku.php?id=technical_support) is appropriate -- we put it to people that they really should be using the up-to-date version of the tool. Granted, regressions can happen as a result of updating to newer versions, but as a practical question, that will be quite rare. But, fine, if there is such a problem, the person should report it to us and we'll deal with it, fine. That's how progress is made.

I mean, really, the whole idea that somebody wants to just use an obsolete version and never upgrade because they're so afraid they'll run into a problem... well, okay, people can do what they want, but to cut oneself off from all the ongoing improvements in the thing and (due to the policy that I set out there) to cut oneself off from getting any technical support, how could that ever be a good deal?

**Comment by hylkevds** [comment]
> I personally see little value in uploading CongoCC to Maven Central, other than to satisfy corporate guidelines mandated in some enviroments. ISTM Maven Central is more for libraries rather than applications.

CongoCC is not just an application, it is part of build envionments, like maven plugins are. Even without a maven plugin to make it easier to integrate CongoCC in builds, the fact that Maven Central guarantees that versions will never change or vanish is quite valuable.

**Reply by revusky** [reply]
>  guarantees that versions will never change or vanish 

Well, I honestly cannot fathom what special "guarantee" you think you're getting. Well, okay, if we upload a file to Maven Central, it will be there and doesn't change, okay, that's true.

But, consider this page: https://github.com/javacc21/javacc21/commits/master/bin/javacc.jar

That is the commit history of the bin/javacc.jar a.k.a. the bootstrap jar that is used to build JavaCC 21. Each one of those binary builds was considered solid enough (correctly or not!) to *re-bootstrap* the build at that point in time. And that was done, by my count, paging through the history, 179 times. So if you want the bootstrap jarfile that was used to build the tool itself from whatever date, you can just download it. Here is the link to the one (just randomly chosen) that was committed on 13 October 2021: https://github.com/javacc21/javacc21/blob/e85d4684795315fbafea10273b9917d6592ffecf/bin/javacc.jar

(That's a download page, but the direct download link for the jarfile is the same except with `raw` the URL instead of `blob`. And that could be used with curl or wget if you were so inclined, I guess.)

As I said separately, there is very rarely any reason to want to download and use any of these. But okay, if you do want to do that, they're all there! And I see no reason at all to think they're going to change or vanish any more than something on Maven Central. It's not something I would lose any sleep over, in any case!

Now, I can grant also that the whole Maven Central setup is more *formalized*, like if you say you're using whatever jarfile from a year ago, you can just put that in your pom.xml or whatever and so on. But.... I don't think that is something that one should really want to do very often anyway. If you really want to rebuild your own project using a much older version of this tool, there's not really much of a problem with that. 

(Note that I'm not totally denying the value of Maven etcetera, but I mean that in our case where we have all these jarfiles that are self-contained. By all means, if these files had in turn other dependencies on other versions of other things... but that is not the case... so I stand by my point that the practical value of all the maven ecology for us is pretty limited!)

All that said, if all this is soooo important to certain corporate people, well, we can see what we can do to make people happy, sure.

**Comment by vsajip** [comment]
For future reference, this is an example of the steps which need to be followed in order to get software onto Maven Central, without using Maven or Gradle:

https://mccue.dev/pages/6-1-22-upload-to-maven-central
